repair/row_level.cc

```diff
@@ -322,11 +329,78 @@ mutation_reader repair_reader::make_reader(
             // So we release the one on _permit so the only one is the one the
             // shard reader will obtain.
             _permit.release_base_resources();
-            return make_filtering_reader(make_multishard_streaming_reader(db, _schema, _permit, _range, compaction_time),
+            co_return make_filtering_reader(make_multishard_streaming_reader(db, _schema, _permit, _range, compaction_time),
                 [&remote_sharder, remote_shard](const dht::decorated_key& k) {
                     return remote_sharder.shard_for_reads(k.token()) == remote_shard;
                 });
         }
+        case read_strategy::multishard_proxy: {
+            _permit.release_base_resources();
+#if 0
+            co_return make_multishard_streaming_reader(db, _schema, _permit, _range, compaction_time);
+#else
+            struct sst_and_shard {
+                unsigned shard;
+                sstring filename;
+            };
+
+            std::vector<foreign_ptr<lw_shared_ptr<utils::chunked_vector<sstables::sstable_files_snapshot>>>> ptrs(smp::count);
+            auto sstable_filenames = co_await db.map_reduce0(
+                [token_range = _token_range, &ptrs, table_id = _schema->id()] (auto& db) -> future<std::vector<sst_and_shard>> {
+                    std::vector<sst_and_shard> res;
+                    auto& t = db.find_column_family(table_id);
+                    auto sstables = co_await t.take_storage_snapshot(token_range);
+                    for (auto& sst : sstables) {
+                        co_await coroutine::maybe_yield();
+                        rlogger.debug("Get sst filename={}", sst.sst->get_filename());
+                        res.push_back(sst_and_shard{this_shard_id(), sst.sst->get_filename()});
+                    }
+                    ptrs[this_shard_id()] = make_foreign(make_lw_shared<utils::chunked_vector<sstables::sstable_files_snapshot>>(std::move(sstables)));
+                    co_return res;
+                },
+                std::vector<sst_and_shard>{},
+                [] (std::vector<sst_and_shard> a, std::vector<sst_and_shard> b) { std::move(b.begin(), b.end(), std::back_inserter(a)); return a;
+            });
+            _sstable_snapshots = std::move(ptrs);
+
```

Comment:
the sstable on remote shard can be deleted by the time map_reduce returns. I suggest you use this sstable::get_open_info(), which can guarantee the sstable can be loaded later, even if it was unlinked. there's a load variant that accepts foreign_sstable_open_info, which is returned by get_open_info(). that way, the load below won't have to read anything from the disk, it will all be copy in memory.